hexkit/providers/s3/provider.py

```diff
@@ -0,0 +1,808 @@
+                Key=object_id,
+            )
+        except botocore.exceptions.ClientError:
+            return False
+
+        return True
+
+    async def _assert_object_exists(self, *, bucket_id: str, object_id: str) -> None:
+        """Checks if the file with specified ID (`object_id`) exists in the bucket with
+        the specified ID (`bucket_id`) and throws an ObjectNotFoundError otherwise.
+        If the bucket does not exist it throws a BucketNotFoundError.
+        """
+
+        # first check if bucket exists:
+        await self._assert_bucket_exists(bucket_id)
+
+        if not await self.does_object_exist(bucket_id=bucket_id, object_id=object_id):
+            raise self.ObjectNotFoundError(bucket_id=bucket_id, object_id=object_id)
+
+    async def _assert_object_not_exists(
+        self, *, bucket_id: str, object_id: str
+    ) -> None:
+        """Checks if the file with specified ID (`object_id`) exists in the bucket with
+        the specified ID (`bucket_id`). If so, it throws an ObjectAlreadyExistsError otherwise.
+        If the bucket does not exist it throws a BucketNotFoundError.
+        """
+
+        # first check if bucket exists:
+        await self._assert_bucket_exists(bucket_id)
+
+        if await self.does_object_exist(bucket_id=bucket_id, object_id=object_id):
+            raise self.ObjectAlreadyExistsError(
+                bucket_id=bucket_id, object_id=object_id
+            )
+
+    async def _get_object_upload_url(
+        self,
+        *,
+        bucket_id: str,
+        object_id: str,
+        expires_after: int = 86400,
+        max_upload_size: Optional[int] = None,
+    ) -> PresignedPostURL:
+        """Generates and returns an HTTP URL to upload a new file object with the given
+        id (`object_id`) to the bucket with the specified id (`bucket_id`).
+        You may also specify a custom expiry duration in seconds (`expires_after`) and
+        a maximum size (bytes) for uploads (`max_upload_size`).
+        """
+
+        await self._assert_object_not_exists(bucket_id=bucket_id, object_id=object_id)
+
+        conditions = (
+            []
+            if max_upload_size is None
+            else [
+                # set upload size limit:
+                ["content-length-range", 0, max_upload_size],
+            ]
+        )
+
+        try:
+            presigned_url = await asyncio.to_thread(
+                self._client.generate_presigned_post,
+                Bucket=bucket_id,
+                Key=object_id,
+                Conditions=conditions,
+                ExpiresIn=expires_after,
+            )
+        except botocore.exceptions.ClientError as error:
+            raise self._translate_s3_client_errors(
+                error, bucket_id=bucket_id, object_id=object_id
+            ) from error
+
+        return PresignedPostURL(
+            url=presigned_url["url"], fields=presigned_url["fields"]
+        )
+
+    async def _list_mulitpart_upload_for_object(
```

Comment:
```suggestion
    async def _list_multipart_upload_for_object(
```
Usages need to be fixed as well